server/channels/store/sqlstore/channel_store.go

```diff
@@ -1922,6 +1922,88 @@ func (s SqlChannelStore) UpdateMemberNotifyProps(channelID, userID string, props
 	return dbMember.ToModel(), err
 }
 
+// PatchMultipleMembersNotifyProps updates the NotifyProps of multiple channel members at once without modifying
+// unspecified fields. Like UpdateMemberNotifyProps, it doesn't modify the LastUpdateAt of affected rows.
+//
+// Note that the returned array may not be in the same order as the provided IDs.
+func (s SqlChannelStore) PatchMultipleMembersNotifyProps(members []*model.ChannelMemberIdentifier, notifyProps map[string]string) ([]*model.ChannelMember, error) {
+	if len(notifyProps) == 0 {
+		return nil, errors.New("PatchMultipleMembersNotifyProps: No notifyProps specified")
+	}
+
+	if err := model.IsChannelMemberNotifyPropsValid(notifyProps, true); err != nil {
+		return nil, err
+	}
+
+	// Make the where clause first since it'll be used multiple times
+	whereClause := sq.Or{}
+	for _, member := range members {
+		whereClause = append(whereClause, sq.And{
+			sq.Eq{"ChannelId": member.ChannelId},
+			sq.Eq{"UserId": member.UserId},
+		})
+	}
+
+	// Update the channel members
+	builder := s.getQueryBuilder().Update("ChannelMembers")
+
+	if s.DriverName() == model.DatabaseDriverPostgres {
+		jsonNotifyProps := string(model.ToJSON(notifyProps))
+		builder = builder.Set("notifyprops", sq.Expr("notifyprops || ?::jsonb", jsonNotifyProps))
+	} else {
+		// Unpack the keys and values to pass to MySQL
+		jsonArgs, jsonSQL := constructMySQLJSONArgs(notifyProps)
+		jsonExpr := sq.Expr(fmt.Sprintf("JSON_SET(NotifyProps, %s)", jsonSQL), jsonArgs...)
+
+		// Example: UPDATE ChannelMembers
+		// SET NotifyProps = JSON_SET(NotifyProps, '$.mark_unread', '"yes"' [, ...])
+		// WHERE ...
+		builder = builder.Set("NotifyProps", jsonExpr)
+	}
+
+	sqlUpdate, args, err := builder.Where(whereClause).ToSql()
+	if err != nil {
+		return nil, errors.Wrap(err, "PatchMultipleMembersNotifyProps_ToSQL")
+	}
+
+	transaction, err := s.GetMasterX().Beginx()
+	if err != nil {
+		return nil, errors.Wrap(err, "begin_transaction")
+	}
+	defer finalizeTransactionX(transaction, &err)
+
+	if result, err := transaction.Exec(sqlUpdate, args...); err != nil {
+		return nil, errors.Wrap(err, "PatchMultipleMembersNotifyProps: Failed to update ChannelMembers")
+	} else if count, _ := result.RowsAffected(); count != int64(len(members)) {
+		return nil, errors.Wrap(err, "PatchMultipleMembersNotifyProps: Unable to update all ChannelMembers, some must not exist")
+	}
```

Comment:
Usually we just check for the error returned from `RowsAffected` ,and that's sufficient.

Reply:
I wanted to confirm that all of the provided ID pairs were actually updated to prevent user error. If we don't count the number of affected rows, then we might end up incorrectly thinking that something was updated when it wasn't. That could also cause later errors if we're expecting the returned array to have the same length as the input array which is why I double checked the number in the SELECT.

In short, I want to keep this check to prevent developer errors, but I'll remove the one from the SELECT query

Reply:
> I wanted to confirm that all of the provided ID pairs were actually updated to prevent user error. If we don't count the number of affected rows, then we might end up incorrectly thinking that something was updated when it wasn't.

Yes, that confirmation can be done by checking the error itself. In short, there shouldn't be a case where err is nil, but still count is different from the number of rows to be updated.

Reply:
I'm not sure if it does actually return an error in that case. From what I can tell, `RowsAffected()` only returns an error if the DB driver isn't able to determine the number of rows affected which shouldn't occur for the Postgres or MySQL drivers we use. If it were to return an error in that case, I'm not sure how it would know the number of rows that we expect to change based on our query since it looks like
```sql
UPDATE
     ChannelMembers
SET ...
WHERE
     (UserId = ? AND ChannelId = ?) OR
     (UserId = ? AND ChannelId = ?) OR
     (UserId = ? AND ChannelId = ?)
```
I don't think there's anything the DB or driver could get from this to tell it that we expect exactly 3 rows to be updated without us checking manually